<center><h1>Applied Data Science Capstone:</h1></center>

<center><h2>Segmenting and clustering Neighborhoods in Toronto</h2></center>

For this assignment, I will be required to explore and cluster the neighborhoods in Toronto


1.   Start by creating a new Notebook for this assignment.
2.   Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M.



## We import the libraries

In [0]:
import numpy as np
import pandas as pd
import urllib

## We import data

In [0]:
page = urllib.request.urlopen("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
content = page.read().decode("UTF-8")

## We create the DataFrame

In [0]:
data = pd.read_html(content, header = 0)[0]
# We change the name of columns
data.columns = ['PostalCode', 'Borough', 'Neighborhood']

In [65]:
data.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

We only process the cells that have an assigned borough. we ignore cells with a borough that is "Not assigned".

In [0]:
# We drop "Not assigned" value in Borough
data = data[data['Borough'] != "Not assigned"]

In [67]:
data.head()

PostalCode           Borough      Neighborhood
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront
5        M5A  Downtown Toronto       Regent Park
6        M6A        North York  Lawrence Heights

More than one neighborhood can exist in one postal code area.

In [0]:
# We list the Postal code:
postal_code = list(data['PostalCode'].unique())
# We create a boucle:
for i in postal_code:
    # We list the value and transform it to a string
    liste = str(list(data[data['PostalCode'] == i]['Neighborhood'].unique()))
    liste = liste.replace("[","")
    liste = liste.replace("]","")
    liste = liste.replace("'","")
    liste = liste.replace('"',"")
    # We list the index
    index = list(data2[data2['PostalCode'] == i].index)
    # We change the Neighborhood
    data.loc[index, "Neighborhood"] = liste

In [69]:
print(data.shape)
data.head()

(211, 3)


PostalCode           Borough                      Neighborhood
2        M3A        North York                         Parkwoods
3        M4A        North York                  Victoria Village
4        M5A  Downtown Toronto         Harbourfront, Regent Park
5        M5A  Downtown Toronto         Harbourfront, Regent Park
6        M6A        North York  Lawrence Heights, Lawrence Manor

We transform "Not assigned" Neighborhood to "Queen's Park"

In [0]:
index = list(data[data['Neighborhood'] == "Not assigned"].index)
data.loc[index, "Neighborhood"] = "Queen's Park"

We clean the dataframe:

In [74]:
data = data.drop_duplicates()
data.head()

PostalCode           Borough                      Neighborhood
2        M3A        North York                         Parkwoods
3        M4A        North York                  Victoria Village
4        M5A  Downtown Toronto         Harbourfront, Regent Park
6        M6A        North York  Lawrence Heights, Lawrence Manor
8        M7A      Queen's Park                      Queen's Park

In [75]:
# We finish to show the number of rows of the DataFrame
data.shape

(103, 3)

There is 103 rows in the DataFrame

# We import latitude and longitude

In [90]:
geospacial = pd.read_csv('http://cocl.us/Geospatial_data')
geospacial.columns = ['PostalCode', 'Latitude', 'Longitude']
geospacial.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

We create a new DataFrame with latitude and longitude

In [91]:
data_total = pd.merge(data, geospacial, on=['PostalCode'], how='inner')
data_total.head(12)

PostalCode           Borough  ...   Latitude  Longitude
0         M3A        North York  ...  43.753259 -79.329656
1         M4A        North York  ...  43.725882 -79.315572
2         M5A  Downtown Toronto  ...  43.654260 -79.360636
3         M6A        North York  ...  43.718518 -79.464763
4         M7A      Queen's Park  ...  43.662301 -79.389494
5         M9A         Etobicoke  ...  43.667856 -79.532242
6         M1B       Scarborough  ...  43.806686 -79.194353
7         M3B        North York  ...  43.745906 -79.352188
8         M4B         East York  ...  43.706397 -79.309937
9         M5B  Downtown Toronto  ...  43.657162 -79.378937
10        M6B        North York  ...  43.709577 -79.445073
11        M9B         Etobicoke  ...  43.650943 -79.554724

[12 rows x 5 columns]